Weather-Widget: replace drop-down selectors with scrollable rows

Sir-Photch opened the pull request:
Had some git-headache. See #250.

Current status:

<details>
<summary>Screenshots</summary>

### Compact:
![weather-widget-compact](https://user-images.githubusercontent.com/47949835/218279767-d3432343-ba89-4542-93c3-c508fb35ae54.png)

### Non-compact:
![weather-widget-expanded](https://user-images.githubusercontent.com/47949835/218279732-7f41c907-a1f5-4ba5-ab1f-47e620e06fc1.png)

</details>

I figured that the time-forecast would also have to be selectable. There is a switch in weather-widget-settings enabling users to choose which layout they would like. Otherwise, this is almost done I think? I thought about some blend-in animation when row-items scroll into view, but I have no idea what I could use to achieve this.

MM2-0 commented:
Current status:

<details>

![Screenshot_20230212-160409_Kvaesitso Debug~2](https://user-images.githubusercontent.com/15646950/218318992-0e804d2d-a143-4169-b1aa-20be61fcb26f.png)

</details>

TODO:

- [ ] Check humidity, precipitation and wind for invalid values and hide icons accordingly

Sir-Photch commented:
Have you thought about adjusting the cloud color dynamically according to the material theme again? For some material themes, the cloud palette does stick out a bit, so I don't think this is a bad idea.

Otherwise, I had the idea to change the city icon on the top left to a ["location icon"](https://fonts.google.com/icons?selected=Material%20Symbols%20Rounded%3Amy_location%3AFILL%400%3Bwght%40400%3BGRAD%400%3Bopsz%4048) if it's the latitude / longitude that is displayed.

Also, how did you generate the code for the icons? Is there some tool that you used? I can't imagine that you copied the values manually. :sweat_smile: I could add the location icon then as well, if it is not already present.

MM2-0 commented:
> Have you thought about adjusting the cloud color dynamically according to the material theme again? For some material themes, the cloud palette does stick out a bit, so I don't think this is a bad idea.

I have tried but I wasn't very satisfied with the results. I might give it another go later.

> Also, how did you generate the code for the icons? Is there some tool that you used?

I use a script, that converts svg paths to compose:

<details>

```js
#!/usr/bin/env node

const parseSVG = require('svg-path-parser')

const d = process.argv[2]

const path = parseSVG(d)

for (const cmd of path) {
    switch (cmd.code) {
        case 'M': console.log(`moveTo(${cmd.x}f, ${cmd.y}f)`)
            break
        case 'm': console.log(`moveToRelative(${cmd.x}f, ${cmd.y}f)`)
            break
        case 'L': console.log(`lineTo(${cmd.x}f, ${cmd.y}f)`)
            break
        case 'l': console.log(`lineToRelative(${cmd.x}f, ${cmd.y}f)`)
            break
        case 'V': console.log(`verticalLineTo(${cmd.y}f)`)
            break
        case 'v': console.log(`verticalLineToRelative(${cmd.y}f)`)
            break
        case 'H': console.log(`horizontalLineTo(${cmd.x}f)`)
            break
        case 'h': console.log(`horizontalLineToRelative(${cmd.x}f)`)
            break
        case 'C': console.log(`curveTo(${cmd.x1}f, ${cmd.y1}f, ${cmd.x2}f, ${cmd.y2}f, ${cmd.x}f, ${cmd.y}f)`)
            break
        case 'c': console.log(`curveToRelative(${cmd.x1}f, ${cmd.y1}f, ${cmd.x2}f, ${cmd.y2}f, ${cmd.x}f, ${cmd.y}f)`)
            break
        case 'S': console.log(`reflectiveCurveTo(${cmd.x2}f, ${cmd.y2}f, ${cmd.x}f, ${cmd.y}f)`)
            break
        case 's': console.log(`reflectiveCurveToRelative(${cmd.x2}f, ${cmd.y2}f, ${cmd.x}f, ${cmd.y}f)`)
            break
        case 'Q': console.log(`quadTo(${cmd.x1}f, ${cmd.y1}f, ${cmd.x}f, ${cmd.y}f)`)
            break
        case 'q': console.log(`quadToRelative(${cmd.x1}f, ${cmd.y1}f, ${cmd.x}f, ${cmd.y}f)`)
            break
        case 'T': console.log(`reflectiveQuadTo(${cmd.x}f, ${cmd.y}f)`)
            break
        case 't': console.log(`reflectiveQuadToRelative(${cmd.x}f, ${cmd.y}f)`)
            break
        case 'A': console.log(`arcTo(${cmd.rx}f, ${cmd.ry}f, ${cmd.xAxisRotation}f, ${cmd.largeArc}, ${cmd.sweep}, ${cmd.x}f, ${cmd.y}f)`)
            break
        case 'a': console.log(`arcToRelative(${cmd.rx}f, ${cmd.ry}f, ${cmd.xAxisRotation}f, ${cmd.largeArc}, ${cmd.sweep}, ${cmd.x}f, ${cmd.y}f)`)
            break
        case 'Z': console.log('close()')
            break
    }
}
```

</details>

> I could add the location icon then as well, if it is not already present.

A similar icon is part of the Material Icons: https://fonts.google.com/icons?selected=Material%20Icons%20Round%3Amy_location%3A so it is already present as part of the material-icons-extended library and can be used with `Icons.Rounded.MyLocation`.

Sir-Photch commented:
> I use a script, that converts svg paths to compose

Thanks for sharing!

> A similar icon is part of the Material Icons

Only now I see that there are two categories: Symbols and Icons. Explains why the other ones are not available.

> I have tried but I wasn't very satisfied with the results.

This also could be something for the user to unset in settings if they don't like it, couldn't it?

I've added checks for valid detail values. Also, I added a regexp for lat/long matching.

Sir-Photch commented on app/ui/src/main/java/de/mm20/launcher2/ui/launcher/widgets/weather/WeatherWidget.kt:
```diff
@@ -430,6 +443,25 @@ fun WeatherDaySelector(
     }
 }
 
+private fun LazyListLayoutInfo.blendIntoViewAlpha(key: Any, degree: Float = 1f): Float =
+    visibleItemsInfo.firstOrNull { it.key == key }?.let {
+
+        val itemStart = it.offset
+        val itemEnd = it.offset + it.size
+
+        val atLeftEnd = itemStart < viewportStartOffset
+        val atRightEnd = itemEnd > viewportEndOffset
+
+        if (!atLeftEnd && !atRightEnd) {
+            return 1f
+        }
+
+        val alpha = 1f - (if (atLeftEnd) viewportStartOffset - itemStart else itemEnd - viewportEndOffset) / it.size.toFloat()
+
+        if (degree != 1f) alpha.pow(degree) else alpha
+
+    } ?: 1f
```
Would this be something to put into core:ktx?

MM2-0 replied:
I don't want Jetpack Compose dependencies in other modules so there is another `ktx` package in `de.mm20.launcher2.ui.ktx`, for extension functions for compose types.

Sir-Photch commented:
I've added blending into / out of view:

<details>
<summary>Screenshot</summary>

![blending](https://user-images.githubusercontent.com/47949835/218863630-4b83faec-9786-4e05-aa74-e8d8e2e6a1aa.png)

</details>

I think it looks pretty nice, before, items were always cut off at the beginning and end of the row. This change makes that less noticeable.
You can adjust how quickly this "blending" is applied by changing the `degree`-parameter of `LazyListLayoutInfo.blendIntoViewAlpha()`, if necessary.

The only thing still bugging me a bit is the provider-info in the top right being somewhat cut off when the card-corner-radius is high. But otherwise, I think this is good to go, isn't it?

MM2-0 commented:
> The only thing still bugging me a bit is the provider-info in the top right being somewhat cut off when the card-corner-radius is high.

True, we should increase the padding a bit.

Sir-Photch commented:
Another thing I've noticed:

<details>
<summary>Screenshot</summary>

![Screenshot_20230214-230603](https://user-images.githubusercontent.com/47949835/218947398-4c8b6a10-64bf-4ace-a642-007aed395dfd.png)

</details>

Apparently, checking  `forecast.precipitation != -1.0` is not enough to determine if there is a valid value in that field. It shows 0mm even though it is raining?

Sir-Photch commented:
It is the same on the latest release:

<details>
<summary>Screenshot</summary>

![Screenshot_20230216-231026~2.png](https://user-images.githubusercontent.com/47949835/219498964-2fcc6033-786e-4313-8926-87d31fff209e.png)

</details>

This seems to be an issue to OpenWeatherMap though, since for other Providers it is fine.

MM2-0 commented:
Maybe it is less than 0.05 mm so it is rounded down to zero.

MM2-0 commented:
Or maybe I made a mistake:

https://github.com/MM2-0/Kvaesitso/blob/ce71784b747b4169c0060dadcecd2a95dd7cc7ef/data/weather/src/main/java/de/mm20/launcher2/weather/openweathermap/OpenWeatherMapApi.kt#L90

(this should be `3h`, not `3d`).
